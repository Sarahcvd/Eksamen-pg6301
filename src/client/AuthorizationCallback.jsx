import * as React from "react";
import { useEffect } from "react";
import { useHistory } from "react-router";

export function AuthorizationCallback() {
  let history = useHistory();
  useEffect(() => {
    const { access_token } = Object.fromEntries(
      new URLSearchParams(window.location.hash.substring(1))
    );
    localStorage.setItem("access_token", access_token);
    history.push("/home");
  }, []);
  return <div>Callback...</div>;
}
