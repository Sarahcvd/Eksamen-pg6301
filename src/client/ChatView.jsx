import React, { useEffect, useState } from "react";

export function ChatView({ username }) {
  const [chatLog, setChatLog] = useState([]);
  const [message, setMessage] = useState("");
  const [ws, setWs] = new useState();

  useEffect(() => {
    const ws = new WebSocket("ws://localhost:3000");
    ws.onmessage = (event) => {
      const { message, id, username } = JSON.parse(event.data);
      setChatLog((chatLog) => [...chatLog, { message, id, username }]);
    };
    ws.onopen = (event) => {
      ws.send(
        JSON.stringify({
          type: "login",
          username,
        })
      );
    };
    setWs(ws);
  }, []);

  function handleSubmitMessage(e) {
    e.preventDefault();
    ws.send(
      JSON.stringify({
        type: "message",
        message: message,
      })
    );
    setMessage("");
  }

  return (
    <div id="chatContainer">
      <header>
        <h1>Chat page</h1>
      </header>
      <main id="chatLog">
        {chatLog.map(({ message, id, username }) => (
          <div key={id}>
            <strong>{username}: </strong>
            {message}
          </div>
        ))}
      </main>
      <footer>
        <form onSubmit={handleSubmitMessage}>
          <input
            type="text"
            autoFocus={true}
            value={message}
            //State is set to the user input inside the forms input field
            onChange={(e) => setMessage(e.target.value)}
          />
          <button>Send</button>
        </form>
      </footer>
    </div>
  );
}
