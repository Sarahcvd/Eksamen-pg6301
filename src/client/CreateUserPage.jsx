import * as React from "react";
import { useState } from "react";
import { InputField } from "./Components/InputField";
import { useHistory } from "react-router";
import { useSubmit } from "./lib/useSubmit";

export function CreateUserPage({ userApi }) {
  const [firstName, setFirstName] = useState("");
  const [lastName, setLastName] = useState("");
  const [email, setEmail] = useState("");
  const history = useHistory();

  const { handleSubmit, submitting, error } = useSubmit(
    async () => {
      await userApi.createUser({ firstName, lastName, email });
    },
    () => history.push("/home")
  );

  return (
    <form onSubmit={handleSubmit}>
      {submitting && <div>Please wait</div>}
      {error && <div>Error: {error.toString()}</div>}
      <h1>Add a new user</h1>
      <InputField
        value={firstName}
        label={"First name"}
        onChangeValue={setFirstName}
      />
      <InputField
        value={lastName}
        label={"Last Name"}
        onChangeValue={setLastName}
      />
      <InputField value={email} label={"Email"} onChangeValue={setEmail} />
      <button disabled={submitting}>Submit</button>
    </form>
  );
}
