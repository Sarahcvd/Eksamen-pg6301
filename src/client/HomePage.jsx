import { Link } from "react-router-dom";
import * as React from "react";

export function HomePage() {
  return (
    <div id={"home-page-container"}>
      <h1>Welcome to this awesome application</h1>
      <h4>Login</h4>
      <li>
        <Link to={"/profile"}>Profile Page</Link>
      </li>
      <h4>User Info</h4>
      <li>
        <Link to={"/create"}>Add a New User</Link>
      </li>
      <li>
        <Link to={"/userInfo"}>See User Info Page</Link>
      </li>
      <h4>Chat</h4>
      <li>
        <Link to={"/chat"}>Chat</Link>
      </li>
    </div>
  );
}
