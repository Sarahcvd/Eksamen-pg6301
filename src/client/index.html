<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8" />
    <title>Hello World!</title>
    <link rel="stylesheet" href="./styles.css">
</head>
<body>
<div id="root"></div>
</body>
<script src="index.jsx"></script>
</html>