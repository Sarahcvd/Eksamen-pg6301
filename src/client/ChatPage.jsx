import * as React from "react";
import { useState } from "react";
import { ChatView } from "./ChatView";

export function ChatPage() {
  const [username, setUsername] = useState();
  if (!username) {
    return <ChatLoginPage onLogin={(username) => setUsername(username)} />;
  }

  return <ChatView username={username} />;
}

function ChatLoginPage({ onLogin }) {
  const [username, setUsername] = useState("");
  function handleSubmit(e) {
    e.preventDefault();
    onLogin(username);
  }
  return (
    <div>
      <h1>Please Enter Username</h1>
      <form onSubmit={handleSubmit}>
        <input
          type="text"
          value={username}
          onChange={(e) => setUsername(e.target.value)}
        />
        <button>Start Chat</button>
      </form>
    </div>
  );
}
