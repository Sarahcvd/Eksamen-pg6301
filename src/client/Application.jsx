import * as React from "react";
import { useState } from "react";
import { Link, Redirect } from "react-router-dom";
import { Route, Switch } from "react-router";
import { ChatPage } from "./ChatPage";
import { LoginPage } from "./LoginPage";
import { EditUserPage } from "./EditUserPage";
import { CreateUserPage } from "./CreateUserPage";
import { HomePage } from "./HomePage";
import { UserInfo } from "./UserInfo";
import { ProfilePage } from "./ProfliePage";
import { fetchJson, postJson } from "./lib/http";
import { LoginCallbackPage } from "./LoginCallbackPage";

export function Application() {
  const userApi = {
    listUsers: async () => await fetchJson("http://localhost:3000/api/users"),
    getUser: async (id) =>
      await fetchJson(`http://localhost:3000/api/users/${id}`, {
        headers: {
          ...(access_token ? { Authorization: `Bearer ${access_token}` } : {}),
        },
      }),
    createUser: async ({ firstName, lastName, email }) => {
      return postJson("http://localhost:3000/api/users", {
        method: "POST",
        json: { firstName, lastName, email },
      });
    },
    updateUser: async (id, { firstName, lastName, email }) =>
      await postJson(`http://localhost:3000/api/users/${id}`, {
        method: "PUT",
        json: { firstName, lastName, email },
      }),
    loginUser: async ({ username, password }) =>
      await postJson("http://localhost:3000/api/login", {
        method: "POST",
        json: { username, password },
      }),
    loadProfile: async () =>
      await fetchJson("http://localhost:3000/api/profile", {
        headers: {
          ...(access_token ? { Authorization: `Bearer ${access_token}` } : {}),
        },
      }),
  };

  const [access_token, setAccess_token] = useState();

  const googleIdentityProvider = {
    discoveryURL:
      "https://accounts.google.com/.well-known/openid-configuration",
    client_id:
      "916384078084-0vdudp3eluljf617umqmtoeuu870iru0.apps.googleusercontent.com",
  };

  return (
    <>
      <nav>
        <Link to={"/home"}>Back to home </Link>
      </nav>
      <main>
        <Switch>
          <Route path={"/profile"}>
            <ProfilePage userApi={userApi} />
          </Route>
          <Route path={"/create"}>
            {!access_token ? (
              <Redirect to={"/"} />
            ) : (
              <CreateUserPage userApi={userApi} />
            )}
          </Route>
          <Route path={"/userInfo"}>
            {!access_token ? (
              <Redirect to={"/"} />
            ) : (
              <UserInfo userApi={userApi} />
            )}
          </Route>
          <Route path={"/chat"}>
            {!access_token ? <Redirect to={"/"} /> : <ChatPage />}
          </Route>
          <Route exact path={"/"}>
            <LoginPage identityProvider={googleIdentityProvider} />
          </Route>
          <Route path={"/login/callback"}>
            <LoginCallbackPage
              identityProvider={googleIdentityProvider}
              onAccessToken={(access_token) => setAccess_token(access_token)}
            />
          </Route>
          <Route path={"/users/:id/edit"}>
            {!access_token ? (
              <Redirect to={"/"} />
            ) : (
              <EditUserPage userApi={userApi} />
            )}
          </Route>
          <Route path={"/home"}>
            <HomePage />
          </Route>
          <Route>
            <h1>Not Found</h1>
          </Route>
        </Switch>
      </main>
    </>
  );
}
