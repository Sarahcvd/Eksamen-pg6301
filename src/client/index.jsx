import React from "react";
import ReactDOM from "react-dom";
import { Application } from "./Application";
import { fetchJson, postJson } from "./lib/http";
import { BrowserRouter } from "react-router-dom";

const userApi = {
  listUsers: async () => await fetchJson("http://localhost:3000/api/users"),
  getUser: async (id) =>
    await fetchJson(`http://localhost:3000/api/users/${id}`),
  createUser: async ({ firstName, lastName, email }) => {
    return postJson("http://localhost:3000/api/users", {
      method: "POST",
      json: { firstName, lastName, email },
    });
  },
  updateUser: async (id, { firstName, lastName, email }) =>
    await postJson(`http://localhost:3000/api/users/${id}`, {
      method: "PUT",
      json: { firstName, lastName, email },
    }),
  loginUser: async ({ username, password }) =>
    await postJson("http://localhost:3000/api/login", {
      method: "POST",
      json: { username, password },
    }),
  showProfile: async () => await fetchJson("http://localhost:3000/api/profile"),
};

ReactDOM.render(
  <BrowserRouter>
    <Application userApi={userApi} />
  </BrowserRouter>,
  document.getElementById("root")
);
