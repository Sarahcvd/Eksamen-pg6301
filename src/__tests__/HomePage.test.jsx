import { act } from "react-dom/test-utils";
import ReactDOM from "react-dom";
import { MemoryRouter } from "react-router";
import * as React from "react";
import { HomePage } from "../client/HomePage";

async function renderForTest(child) {
  const container = document.createElement("div");
  await act(async () => {
    await ReactDOM.render(<MemoryRouter>{child}</MemoryRouter>, container);
  });
  return container;
}

describe("home page", () => {
  it("can show home page", async () => {
    const container = await renderForTest(<HomePage />);
    expect(container.innerHTML).toMatchSnapshot();
    expect(container.querySelector("h1").textContent).toEqual(
      "Welcome to this awesome application"
    );
  });
});
