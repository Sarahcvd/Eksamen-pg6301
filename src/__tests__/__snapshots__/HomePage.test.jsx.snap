// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`home page can show home page 1`] = `"<div id=\\"home-page-container\\"><h1>Welcome to this awesome application</h1><li><a href=\\"/profile\\">Profile Page</a></li><h4>User Info</h4><li><a href=\\"/create\\">Add a New User</a></li><li><a href=\\"/userInfo\\">See User Info</a></li><h4>Chat</h4><li><a href=\\"/chat\\">Chat</a></li></div>"`;
